File: taller_hilos/ejercicio9/ejercicio9.py
import threading
import time

# Ejercicio 9: Comparar ejecución secuencial vs concurrente
#
# Secuencial: start() + join() en el mismo ciclo → cada hilo espera al anterior.
# Concurrente: todos los start() primero, todos los join() después → se solapan.
# La diferencia de tiempo comprueba empíricamente el beneficio de la concurrencia.

def trabajo(nombre):
    for i in range(1, 4):
        print(f"{nombre} procesando item {i}")
        # Simula trabajo que tarda tiempo (p.ej. I/O, red).
        time.sleep(0.5)

# --- Ejecución SECUENCIAL ---
print("=== Ejecución SECUENCIAL ===")
inicio = time.time()

for nombre in ["Tarea-1", "Tarea-2"]:
    t = threading.Thread(target=trabajo, args=(nombre,))
    t.start()
    # join() aquí mismo hace que el bucle espere a que este hilo acabe
    # antes de crear el siguiente → ejecución uno tras otro.
    t.join()

print(f"Tiempo secuencial: {(time.time() - inicio):.2f} s\n")

# --- Ejecución CONCURRENTE ---
print("=== Ejecución CONCURRENTE ===")
inicio = time.time()

# Creamos y arrancamos todos los hilos sin esperar a ninguno.
hilos = [threading.Thread(target=trabajo, args=(f"Tarea-{i}",)) for i in range(1, 3)]
for h in hilos:
    h.start()

# Solo después de arrancar todos esperamos a que terminen.
for h in hilos:
    h.join()

print(f"Tiempo concurrente: {(time.time() - inicio):.2f} s")
# El tiempo concurrente debería ser aproximadamente la mitad del secuencial.

File: taller_hilos/ejercicio9/Ejercicio9.java
// Ejercicio 9: Comparar ejecución secuencial vs concurrente
//
// La clave está en el uso de join():
//   - Secuencial: arrancamos T1, esperamos (join), arrancamos T2, esperamos.
//     El tiempo total es la suma de los tiempos individuales.
//   - Concurrente: arrancamos T1 y T2, luego esperamos a los dos.
//     El tiempo total es el del hilo más lento, no la suma.

class Trabajo implements Runnable {
    private String nombre;

    Trabajo(String nombre) {
        this.nombre = nombre;
    }

    public void run() {
        // Cada "item" simula una operación que tarda 500 ms (e.g., llamada a red).
        for (int i = 1; i <= 3; i++) {
            System.out.println(nombre + " procesando item " + i);
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

public class Ejercicio9 {
    public static void main(String[] args) throws InterruptedException {

        // --- Ejecución SECUENCIAL ---
        System.out.println("=== Ejecución SECUENCIAL ===");
        long inicio = System.currentTimeMillis();

        Thread t1 = new Thread(new Trabajo("Tarea-1"));
        Thread t2 = new Thread(new Trabajo("Tarea-2"));

        // join() después de cada start() obliga a que T1 termine
        // antes de lanzar T2 → comportamiento secuencial.
        t1.start(); t1.join();
        t2.start(); t2.join();

        long tiempoSecuencial = System.currentTimeMillis() - inicio;
        System.out.println("Tiempo secuencial: " + tiempoSecuencial + " ms\n");

        // --- Ejecución CONCURRENTE ---
        System.out.println("=== Ejecución CONCURRENTE ===");
        inicio = System.currentTimeMillis();

        Thread c1 = new Thread(new Trabajo("Tarea-1"));
        Thread c2 = new Thread(new Trabajo("Tarea-2"));

        // Arrancamos ambos antes de hacer join() a cualquiera.
        // Los dos hilos trabajan al mismo tiempo.
        c1.start();
        c2.start();
        c1.join(); // Esperamos a que c1 termine...
        c2.join(); // ...luego a que c2 termine (probablemente ya acabó).

        long tiempoConcurrente = System.currentTimeMillis() - inicio;
        System.out.println("Tiempo concurrente: " + tiempoConcurrente + " ms");
        // El tiempo concurrente debería ser ~la mitad del secuencial.
    }
}
